### PR TITLE
Add new medical history options 

### PR DESCRIPTION
**Pull request by AshrafMd-1**
## Proposed Changes

- Added new medical history options 
- Altered the disease table to include `status` `type` & `duration`

### Associated Issue

- https://github.com/coronasafe/care_fe/pull/7852 - Frontend
- Will handle the backend 

### Architecture changes

- Remove this section if not used

## Merge Checklist

- [ ] Tests added/fixed
- [ ] Update docs in `/docs`
- [ ] Linting Complete
- [ ] Any other necessary step

_*Only PR's with test cases included and passing lint and test pipelines will be reviewed*_

@coronasafe/care-backend-maintainers @coronasafe/care-backend-admins

### CODE REVIEW

File: care/facility/api/serializers/patient.py [hunk 1]
```diff
@@ -159,6 +159,9 @@ class PatientDetailSerializer(PatientListSerializer):
     class MedicalHistorySerializer(serializers.Serializer):
         disease = ChoiceField(choices=DISEASE_CHOICES)
         details = serializers.CharField(required=False, allow_blank=True)
+        status = serializers.CharField(required=False, allow_blank=True)
+        duration = serializers.CharField(required=False, allow_blank=True)
+        type  = serializers.CharField(required=False, allow_blank=True)
```
Comment (sainak) [inline care/facility/api/serializers/patient.py]: you can add validation for type based on the disease

File: care/facility/models/patient_base.py [hunk 2]
```diff
@@ -53,10 +53,17 @@ def reverse_choices(choices):
...
+    "Asthma": 6,
+    "COPD": 7,
+    "Bronchitis": 8,
+    "Chronic Neurological Or Neuromuscular Disease": 9,
+    "Immunocompromised Condition": 10,
+    "Liver Disease": 11,
+    "TB": 12,
+    "Other Chronic Lung Diseases": 13,
+    "Cancer": 14,
+    "OTHER": 15,
```
Comment (rithviknishad) [inline care/facility/models/patient_base.py]: Migration of existing records would be incorrect this way. For example, if a patient had a disease "OTHER" selected, it'd be "Bronchitis" once this migration runs, which is incorrect. 

Keep the same numeric value for things that are only just renamed. All new diseases that are being introduced to be added with a new numeric value not used before.

File: care/facility/models/patient.py [hunk 3]
```diff
@@ -677,6 +677,9 @@ class Disease(models.Model):
     )
     disease = models.IntegerField(choices=DISEASE_CHOICES)
     details = models.TextField(blank=True, null=True)
+    status=models.TextField(blank=True, null=True)
+    duration = models.IntegerField(blank=True, null=True)
+    type =models.TextField(blank=True, null=True)
```
Comment (rithviknishad) [inline care/facility/models/patient.py]: ensure code is formatted correctly.

**Review by rithviknishad — CHANGES_REQUESTED** [review]

File: care/facility/api/serializers/patient.py (hunk 1 above)
Comment (AshrafMd-1) [reply care/facility/api/serializers/patient.py]: @sainak Can you give me a pointer on how/where to add validation

**Comment by AshrafMd-1** [comment]
@rithviknishad also is it required to push the migrations?

**Comment by rithviknishad** [comment]
Yup, otherwise the test workflows would fail.

**Comment by AshrafMd-1** [comment]
@rithviknishad can u check once

File: care/facility/api/serializers/patient.py (hunk 1 above)
Comment (rithviknishad) [reply care/facility/api/serializers/patient.py]: @AshrafMd-1 you can extend `validate` method provided by the Serializer class.

Refer: https://www.django-rest-framework.org/api-guide/serializers/#object-level-validation

**Review by rithviknishad — COMMENTED** [review]
Other than https://github.com/coronasafe/care/pull/2190#discussion_r1610985471 and adding tests for those, LGTM.

File: care/facility/models/patient_base.py [hunk 4]
```diff
@@ -53,11 +53,19 @@ def reverse_choices(choices):
     "Diabetes": 2,
     "Heart Disease": 3,
     "HyperTension": 4,
-    "Kidney Diseases": 5,
-    "Lung Diseases/Asthma": 6,
+    "Chronic Renal Disease": 5,
+    "Asthma": 6,
```
Comment (sainak) [inline care/facility/models/patient_base.py]: can we change `Lung Diseases/Asthma` to `Other Chronic Lung Diseases` and make `Asthma` the new separate option?

cc: @aparnacoronasafe

**Comment by AshrafMd-1** [comment]
@rithviknishad @sainak  
Can you take a look at the validation

**Review by rithviknishad — CHANGES_REQUESTED** [review]
LGTM, tests needed for the validations

**Comment by AshrafMd-1** [comment]
@rithviknishad where to add tests ?

**Comment by rithviknishad** [comment]
> @rithviknishad where to add tests ?

https://github.com/coronasafe/care/blob/develop/care/facility/models/tests/test_patient.py#L11

File: care/facility/api/serializers/patient.py [hunk 5]
```diff
@@ -281,6 +285,44 @@ def validate(self, attrs):
             if validated.get("vaccine_name") is None:
                 raise serializers.ValidationError("Vaccine name cannot be null")
 
+        if validated.get('medical_history'):
+            medical_history = validated['medical_history']
```
Comment (sainak) [inline care/facility/api/serializers/patient.py]:
```suggestion
        if medical_history := validated.get('medical_history'):
```

File: care/facility/api/serializers/patient.py [hunk 6]
```diff
@@ -281,6 +285,44 @@ def validate(self, attrs):
...
+
+                        invalid_gender_type = (
+                            (gender == 1 and CANCER_TYPE[cancer_type] in {"1", "5"}) or
+                            (gender == 2 and CANCER_TYPE[cancer_type] == "8")
+                        )
+                        if invalid_gender_type:
+                            raise serializers.ValidationError("Invalid cancer type for specified gender")
+
+                # Handle TB cases
+                if disease_id == 14:
```
Comment (sainak) [inline care/facility/api/serializers/patient.py]:
```suggestion
                elif disease_id == 14:
```

**Comment by AshrafMd-1** [comment]
@rithviknishad are these valid tests
LIke when I run tests it says ok

**Review by rithviknishad — APPROVED** [review]

**Comment by nihal467** [comment]
@AshrafMd-1 can you fix the lint and test failure 

CC: @rithviknishad @AshrafMd-1

**Comment by AshrafMd-1** [comment]
> @AshrafMd-1 can you fix the lint and test failure
> 
> CC: @rithviknishad @AshrafMd-1

@nihal467 can you check it now

**Review by khavinshankar — APPROVED** [review]

**Comment by khavinshankar** [comment]
@AshrafMd-1 can you fix the linting errors

**Comment by AshrafMd-1** [comment]
@khavinshankar can you review it once

**Review by rithviknishad — APPROVED** [review]

**Comment by nihal467** [comment]
LGTM

File: care/facility/api/serializers/patient.py [hunk 7]
```diff
@@ -272,6 +276,49 @@ def validate(self, attrs):
...
+                    cancer_type = disease.get("type")
+                    if cancer_type:
+                        gender = validated.get("gender")
+                        if cancer_type not in CANCER_TYPE:
+                            raise serializers.ValidationError("Invalid cancer type")
+
+                        invalid_gender_type = (
+                            gender == 1 and CANCER_TYPE[cancer_type] in {"1", "5"}
+                        ) or (gender == 2 and CANCER_TYPE[cancer_type] == "8")
+                        if invalid_gender_type:
```
Comment (sainak) [inline care/facility/api/serializers/patient.py]:
```suggestion

                        if (
                            gender == 1 and CANCER_TYPE[cancer_type] in {"1", "5"}
                        ) or (gender == 2 and CANCER_TYPE[cancer_type] == "8"):
```

File: care/facility/api/serializers/patient.py [hunk 8]
```diff
@@ -272,6 +276,49 @@ def validate(self, attrs):
...
+                            )
+
+                # Handle TB cases
+                if disease_id == 14:
+                    status = disease.get("status")
+                    if status and status not in {"Active", "Old"}:
+                        raise serializers.ValidationError("Invalid TB status")
+
+                    duration = disease.get("duration")
+                    if duration:
```
Comment (sainak) [inline care/facility/api/serializers/patient.py]:
```suggestion
                    if duration := disease.get("duration"):
```

File: care/facility/migrations/0439_disease_duration_disease_status_disease_type_and_more.py
Comment (sainak) [inline care/facility/migrations/0439_disease_duration_disease_status_disease_type_and_more.py:1]: delete migrations and rerun makemigrate

**Review by sainak — CHANGES_REQUESTED** [review]

**Comment by AshrafMd-1** [comment]
@sainak can you check once

**Review by sainak — APPROVED** [review]

**Comment by AshrafMd-1** [comment]
@rithviknishad the codecov check says it failed 

is it required to reach the required percentage ?

**Comment by rithviknishad** [comment]
@AshrafMd-1 You can make use of code coverage inline comments to improve the test coverage

image

**Comment by nihal467** [comment]
@AshrafMd-1 can you look into the comment by rithvik

**Comment by AshrafMd-1** [comment]
@nihal467 can you run the tests once again 

CC: @rithviknishad

**Comment by nihal467** [comment]
@AshrafMd-1 the tests are failing

**Comment by AshrafMd-1** [comment]
> @AshrafMd-1 the tests are failing
> 

I will push a pr by tmr

**Comment by AshrafMd-1** [comment]
@nihal467 i fixed the migration issue can you rerun the tests

**Comment by nihal467** [comment]
https://rebuildearth.slack.com/archives/C06KGMUAB0V/p1728116762128059?thread_ts=1728075807.176319&cid=C06KGMUAB0V

Closing this PR as the entire medical history will be re-built after verifying proper standards and included in the health details API.

**Pull request closed without merging**